File: DataBindingStudy/app/src/main/java/com/jie/databinding/ObservableFieldActivity.java
package com.jie.databinding;

import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;

import android.os.Bundle;

import com.jie.databinding.databinding.ActivityObservableFieldBinding;
import com.jie.databinding.model.ObservableGoods;

import java.lang.ref.WeakReference;
import java.util.Random;

public class ObservableFieldActivity extends AppCompatActivity {
    private ObservableGoods mGoods;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ActivityObservableFieldBinding fieldBinding = DataBindingUtil.setContentView(this, R.layout.activity_observable_field);
        mGoods = new ObservableGoods("dogeCoin", "wa, wonderfull coin", 0.15F);
        fieldBinding.setGoods(mGoods);
        fieldBinding.setDataHandler(new DataHandler(this));
    }

    public ObservableGoods getGoods() {
        return mGoods;
    }

    public static class DataHandler {
        private WeakReference<ObservableFieldActivity> mReference;
        public DataHandler(ObservableFieldActivity activity) {
            mReference = new WeakReference<>(activity);
        }

        public void changeGoodsName() {
            if (mReference != null) {
                ObservableFieldActivity activity = mReference.get();
                ObservableGoods goods = activity.getGoods();
                goods.setName("DogeCoin super" + new Random().nextInt(100));
            }
        }

        public void changeGoodsDetails() {
            if (mReference != null) {
                ObservableFieldActivity activity = mReference.get();
                ObservableGoods goods = activity.getGoods();
                goods.setDetails("In 2022, Elon Musk said: DogeCoin can by some Tesla Belonings." + new Random().nextInt(100));
            }
        }
    }
}

File: DataBindingStudy/app/src/main/java/com/jie/databinding/ui/activity/ObservableFieldActivity.kt
package com.jie.databinding.ui.activity

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.databinding.DataBindingUtil
import com.jie.databinding.R
import com.jie.databinding.databinding.ActivityObservableFieldBinding
import com.jie.databinding.model.ObservableGoodsEntity
import java.lang.ref.WeakReference

class ObservableFieldActivity : AppCompatActivity() {
    private lateinit var goods: ObservableGoodsEntity
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val fieldBinding = DataBindingUtil.setContentView<ActivityObservableFieldBinding>(this, R.layout.activity_observable_field)
        goods = ObservableGoodsEntity("dogeCoin yeah", "wa, wonderfull coin", 0.15F)
        fieldBinding.goods = goods
        fieldBinding.dataHandler = DataHandler(this)
    }

    inner class DataHandler(activity: ObservableFieldActivity) {
        private var mReference: WeakReference<ObservableFieldActivity>? = null
        init {
            mReference = WeakReference(activity)
        }

        fun changeGoodsName() {
            mReference?.let {
                it.get()?.goods?.name?.set("DogeCoin super ${(0..100).random()}")
            }
        }

        fun changeGoodsDetails() {
            mReference?.get()?.goods?.details?.set("In 2022, Elon Musk said: DogeCoin can by some Tesla Belonings. ${(0..100).random()}")
        }
    }
}